src/main/java/com/metamx/common/parsers/JSONPathParser.java

```diff
@@ -0,0 +1,272 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.metamx.common.parsers;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.base.Charsets;
+import com.jayway.jsonpath.Configuration;
+import com.jayway.jsonpath.JsonPath;
+import com.jayway.jsonpath.Option;
+import com.metamx.common.Pair;
+import com.metamx.common.StringUtils;
+
+import java.math.BigInteger;
+import java.nio.charset.CharsetEncoder;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * JSON parser class that uses the JsonPath library to access fields via path expressions.
+ */
+public class JSONPathParser implements Parser<String, Object>
+{
+  private final Map<String, Pair<FieldType, JsonPath>> fieldPathMap;
+  private final List<FieldSpec> fieldSpecs;
+  private final boolean useFieldDiscovery;
+  private final ObjectMapper mapper;
```

Comment:
You may find considerable CPU savings from keeping a static ObjectMapper around for the default case rather than creating a new ObjectMapper in the constructor. ObjectMappers have significant startup costs associated with reflection. They're thread-safe, so re-use of a singleton in concurrent scenarios is fine.

So imagine a static field 

`private static final ObjectMapper DEFAULT_MAPPER = new ObjectMapper();`

then the constructor below on line 63 can read
`this.mapper = mapper == null ? DEFAULT_MAPPER : mapper;`
or (since you're willing to use Guava)
`this.mapper = Objects.firstNonNull(mapper, DEFAULT_MAPPER);`